parse/BuildingsParser.cpp

```diff
@@ -48,21 +40,23 @@ namespace {
             qi::_b_type _b;
             qi::_c_type _c;
             qi::_d_type _d;
+            qi::_pass_type _pass;
             qi::_r1_type _r1;
             qi::eps_type eps;
 
             const parse::lexer& tok = parse::lexer::instance();
 
             building_type
                 =   tok.BuildingType_
-                >   parse::detail::label(Name_token)                > tok.string        [ _a = _1 ]
+                >   parse::detail::label(Name_token)
+                >   tok.string        [ _pass = is_unique_(_r1, "BuildingType", _1), _a = _1 ]
```

Comment:
The `_pass_type` parameter is the third parameter to the `boost::spirit` semantic actions.  Setting it to false causes the parse to fail at that point.

This line calls `is_unique_()` to check if the name is already in the map of existing building types.  If the name is not unique, that will set `_pass` false and cause the parse to fail and the error message generated by `boost::spirit` will point at the name token to indicate where the parse failed.